tslearn/early_classification.py

```diff
@@ -0,0 +1,312 @@
+    making problem. 2015 Conference paper
+
+
+    """
+
+    def __init__(
+        self,
+        n_clusters,
+        base_classifier,
+        minimum_time_stamp,
+        lamb,
+        cost_time_parameter,
+    ):
+        super(NonMyopicEarlyClassification, self).__init__()
+        self.base_classifier = base_classifier
+        self.n_clusters = n_clusters
+        self.minimum_time_stamp = minimum_time_stamp
+        self.lamb = lamb
+        self.cost_time_parameter = cost_time_parameter
+
+    def fit(self, X, y):
+        """
+        This function fits classifiers that are currently multilayer perceptrons to a training set of time series and
+        associated classes. A classifier is fit for each time stamp above a minimum time stamp that is an attribute of
+        the class of the model. Then some probabilities are computed using clusters already computed.
+        This function should be divided among 2 or 3 functions.
+
+        Parameters
+        ----------
+        X: Array-like
+            a dataset of time series
+        y: vector
+            the associated classes of the series from X_train
+        """
+        y_classes = np.unique(y)
+        self.labels_ = sorted(set(y_classes))
+        y_classes_indices = [self.labels_.index(yi) for yi in y_classes]
+
+        for idx, current_classe in enumerate(y_classes):
+            y[y == current_classe] = y_classes_indices[idx]
+
+        self.cluster_ = TimeSeriesKMeans(n_clusters=self.n_clusters)
+        self.classifiers_ = {
+            t: clone(self.base_classifier)
+            for t in range(self.minimum_time_stamp, X.shape[1] + 1)
+        }
+        self.pyhatyck_ = {}
+        self.indice_ck_ = []
+        self.__n_classes_ = len(y_classes_indices)
+        self.__len_X_ = X.shape[1]
+        c_k = self.cluster_.fit_predict((to_time_series_dataset(X)))
+        X_and_cluster = np.concatenate((X, c_k[:, np.newaxis]), axis=1)
+        X1, X2, y1, y2 = train_test_split(X_and_cluster, y, test_size=0.5)
+        X1 = X1[:, :-1]
+        c_k2 = X2[:, -1]
+        X2 = X2[:, :-1]
+        vector_of_ones = np.ones((len(X[:]),))
+        self.pyck_ = coo_matrix(
+            (vector_of_ones, (y, c_k)),
+            shape=(self.__n_classes_, self.n_clusters),
+        ).toarray()
+        for k in range(0, self.n_clusters):
+            self.pyhatyck_["pyhatycks{0}".format(k)] = []
+            self.indice_ck_.append(np.where(c_k == k))
+            current_sum = self.pyck_[:, k].sum()
+            for current_classe in range(0, self.__n_classes_):
+                self.pyck_[current_classe, k] = (
+                    self.pyck_[current_classe, k] / current_sum
+                )
+
+        for t in range(self.minimum_time_stamp, X.shape[1] + 1):
+            self.classifiers_[t].fit(X1[:, :t], y1)
+            for k in range(0, self.n_clusters):
+                index_cluster = np.where(c_k2 == k)
```

Comment:
The fact that you sometimes have to make this a tuple, sometimes you need to squeeze, means that this indexing does not seem appropriate.
It looks to me that the following should work (assuming `c_k2` is of shape `(n, )`), but I have not tested it:

```python
indices = (c_k2 == k)
[...]
X2_cluster = X2[indices, :t]
y2_cluster = y2[indices]
```